controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1420,6 +1486,89 @@ private TxnStatus handleDataNotFoundException(Throwable ex) {
+                    } else {
+                        // sanity check and update
+                        if (record.getObject().getTimestamp() > timestamp) {
+                            // existing time is already ahead of new time
+                            return CompletableFuture.completedFuture(WriterTimestampResponse.INVALID_TIME);
+                        }
+                        if (!compareWriterPositions(record.getObject().getPosition(), newPosition)) {
+                            // existing position is already ahead of new position
+                            return CompletableFuture.completedFuture(WriterTimestampResponse.INVALID_POSITION);
+                        }
+                        
+                        // its a valid mark, update it
+                        return updateWriterMarkRecord(writer, timestamp, newPosition, record.getVersion())
+                                .thenApply(v -> WriterTimestampResponse.SUCCESS);
+                    }
+                });
+    }
+
+
+    /**
+     * Compares two given positions and returns true if position 2 is ahead of position 1. 
+     * 
+     * Note: This is not absolutely correct implementation because that would be a very costly computation. 
+     * This method simply checks if position 2 has a higher max segment number than position 1. 
+     * If those are equal, then it simply checks the offsets in each segment in position 2 to be ahead of same segment in 
+     * position 1. 
+     * It doesnt deal with complex cases where in a position 1 there could be both predecessor and successor while position 2 
+     * only contains successor. There can be many such complex checks which will require looking up ranges of segments involved
+     * which we will avoid in this method. 
+     * 
+     * @param position1 position 1 
+     * @param position2 position 2
+     * @return return true if position 2 is ahead of/or equal to position 1. false otherwise. 
+     */
+    @VisibleForTesting
+    boolean compareWriterPositions(Map<Long, Long> position1, Map<Long, Long> position2) {
+        ArrayList<Map.Entry<Long, Long>> sortedPosition1 = new ArrayList<>(position1.entrySet());
+        sortedPosition1.sort(Collections.reverseOrder(Comparator.comparingLong(Map.Entry::getKey)));
+        ArrayList<Map.Entry<Long, Long>> sortedPosition2 = new ArrayList<>(position2.entrySet());
+        sortedPosition2.sort(Collections.reverseOrder(Comparator.comparingLong(Map.Entry::getKey)));
+
+        long maxInPos2Only = position2.keySet().stream().filter(position1::containsKey).max(Long::compare).orElse(Long.MIN_VALUE);
+        long maxInPos1Only = position1.keySet().stream().filter(position2::containsKey).max(Long::compare).orElse(Long.MIN_VALUE);
+        
+        boolean compareMaxes = maxInPos2Only >= maxInPos1Only;
+        
+        boolean compareOverlaps = position2.entrySet().stream().filter(x -> position1.containsKey(x.getKey()))
```

Comment:
Do this computation lazily, only if `compareMaxes` is false.

Reply:
Also you are double-getting from `position1`. Once in `filter` and once in `allMatch`.

Reply:
Actually, this was a miss on my part. I had a simple check in the first commit. Then i tried improving the comparison check for two positions and realized that without retrieving metadata information about ranges corresponding to segments, it would not be possible to draw any stronger conclusions so i reverted my changes but did not clean it properly. I dont need sorted position sets. I just need to find max and compare maxes.